File: Vista/VentaPanel.py
from PyQt6.QtWidgets import (QWidget, QVBoxLayout, QHBoxLayout, QLabel, QLineEdit, 
                             QPushButton, QTableWidget, QTableWidgetItem, QHeaderView, 
                             QComboBox, QMessageBox, QScrollArea, QFrame)
from PyQt6.QtCore import Qt, QSize
from PyQt6.QtGui import QFont, QColor, QPixmap
from Vista.Estilos import Estilos
import os

class VentaPanel(QWidget):
    def __init__(self, ctrl, config, cajero):
        super().__init__()
        self.ctrl = ctrl
        self.config = config
        self.cajero = cajero
        self.ctrl.setPanel(self)
        self.actualizando = False
        self.construir()

    def construir(self):
        self.setStyleSheet(f"background-color: {Estilos.BG_CLARO.name()};")
        layout = QHBoxLayout(self)
        layout.setContentsMargins(16, 14, 16, 14)

        # Contenido Izquierdo
        panelIzq = QWidget()
        layoutIzq = QVBoxLayout(panelIzq)
        layoutIzq.setContentsMargins(0, 0, 0, 0)

        lblBusq = QLabel("BUSCAR PRODUCTO (nombre o SKU):")
        lblBusq.setFont(Estilos.FUENTE_XS)
        lblBusq.setStyleSheet(f"color: {Estilos.TEXTO_TENUE.name()}; border: none;")
        layoutIzq.addWidget(lblBusq)

        self.txtBusqueda = QLineEdit()
        self.txtBusqueda.setFont(Estilos.FUENTE_SUBTITULO)
        self.txtBusqueda.setStyleSheet(f"""
            QLineEdit {{
                background-color: {Estilos.BG_BLANCO.name()};
                color: {Estilos.TEXTO_PRINCIPAL.name()};
                border: 2px solid {Estilos.BORDE.name()};
                padding: 8px 12px;
            }}
        """)
        self.txtBusqueda.textChanged.connect(self.actualizarSugerencias)
        layoutIzq.addWidget(self.txtBusqueda)

        self.sugerenciasArea = QScrollArea()
        self.sugerenciasArea.setWidgetResizable(True)
        self.sugerenciasArea.setMaximumHeight(150)
        self.sugerenciasArea.setVisible(False)
        self.sugerenciasWidget = QWidget()
        self.sugerenciasLayout = QVBoxLayout(self.sugerenciasWidget)
        self.sugerenciasLayout.setContentsMargins(0, 0, 0, 0)
        self.sugerenciasLayout.setSpacing(0)
        self.sugerenciasArea.setWidget(self.sugerenciasWidget)
        layoutIzq.addWidget(self.sugerenciasArea)

        lblDetalle = QLabel("DETALLE DE VENTA")
        lblDetalle.setFont(Estilos.FUENTE_XS)
        lblDetalle.setStyleSheet(f"color: {Estilos.TEXTO_TENUE.name()}; border: none;")
        layoutIzq.addWidget(lblDetalle)

        self.tablaCarrito = QTableWidget(0, 6)
        self.tablaCarrito.setHorizontalHeaderLabels(["ID", "Producto", "Precio Unit", "Cant", "Subtotal", "Quitar"])
        self.tablaCarrito.setFont(Estilos.FUENTE_NORMAL)
        self.tablaCarrito.horizontalHeader().setFont(Estilos.FUENTE_XS)
        self.tablaCarrito.horizontalHeader().setStyleSheet(f"background-color: {Estilos.BG_ZINC_100.name()}; color: {Estilos.TEXTO_PRINCIPAL.name()};")
        self.tablaCarrito.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()}; color: {Estilos.TEXTO_PRINCIPAL.name()}; border: 1px solid {Estilos.BORDE.name()};")
        self.tablaCarrito.setColumnHidden(0, True)
        self.tablaCarrito.horizontalHeader().setSectionResizeMode(1, QHeaderView.ResizeMode.Stretch)
        self.tablaCarrito.setColumnWidth(2, 120)
        self.tablaCarrito.setColumnWidth(3, 80)
        self.tablaCarrito.setColumnWidth(4, 110)
        self.tablaCarrito.setColumnWidth(5, 100)
        self.tablaCarrito.cellChanged.connect(self.onCeldaCambiada)
        
        layoutIzq.addWidget(self.tablaCarrito)
        layout.addWidget(panelIzq, stretch=3)

        # Panel Cobro
        panelCobro = QWidget()
        panelCobro.setStyleSheet(f"background-color: {Estilos.BG_OSCURO.name()}; border: 1px solid #3F3F46;")
        panelCobro.setFixedWidth(250)
        layoutCobro = QVBoxLayout(panelCobro)
        layoutCobro.setContentsMargins(16, 16, 16, 16)

        lblResumen = QLabel("RESUMEN")
        lblResumen.setFont(Estilos.FUENTE_XS)
        lblResumen.setStyleSheet("color: #818CF8; border: none;")
        layoutCobro.addWidget(lblResumen)

        self.lblTotal = QLabel("$0.00")
        self.lblTotal.setFont(QFont("SansSerif", 38, QFont.Weight.Bold))
        self.lblTotal.setStyleSheet("color: #A5B4FC; border: none;")
        layoutCobro.addWidget(self.lblTotal)

        lblDescuento = QLabel("DESCUENTO ($ o %)")
        lblDescuento.setFont(Estilos.FUENTE_XS)
        lblDescuento.setStyleSheet("color: #71717A; border: none;")
        layoutCobro.addWidget(lblDescuento)

        self.txtDescuento = QLineEdit("0.00")
        self.txtDescuento.setFont(Estilos.FUENTE_BOLD)
        self.txtDescuento.setStyleSheet("""
            QLineEdit {
                background-color: #27272A;
                color: #34D399;
                border: 2px solid #3F3F46;
                padding: 6px 10px;
            }
        """)
        self.txtDescuento.textChanged.connect(self.aplicarDescuento)
        layoutCobro.addWidget(self.txtDescuento)

        lblMetodo = QLabel("METODO DE PAGO")
        lblMetodo.setFont(Estilos.FUENTE_XS)
        lblMetodo.setStyleSheet("color: #71717A; border: none;")
        layoutCobro.addWidget(lblMetodo)

        self.cmbMetodo = QComboBox()
        self.cmbMetodo.addItems(["Efectivo", "Tarjeta"])
        self.cmbMetodo.setFont(Estilos.FUENTE_BOLD)
        self.cmbMetodo.currentIndexChanged.connect(self.actualizarVisibilidadEfectivo)
        layoutCobro.addWidget(self.cmbMetodo)

        self.lblRecibido = QLabel("IMPORTE RECIBIDO")
        self.lblRecibido.setFont(Estilos.FUENTE_XS)
        self.lblRecibido.setStyleSheet("color: #71717A; border: none;")
        layoutCobro.addWidget(self.lblRecibido)

        self.txtRecibido = QLineEdit("0.00")
        self.txtRecibido.setFont(QFont("SansSerif", 22, QFont.Weight.Bold))
        self.txtRecibido.setStyleSheet("""
            QLineEdit {
                background-color: #27272A;
                color: white;
                border: 2px solid #3F3F46;
                padding: 8px 12px;
            }
        """)
        self.txtRecibido.textChanged.connect(self.actualizarCambio)
        layoutCobro.addWidget(self.txtRecibido)

        self.lblCambio = QLabel("Cambio: $0.00")
        self.lblCambio.setFont(Estilos.FUENTE_BOLD)
        self.lblCambio.setStyleSheet("color: #34D399; border: none;")
        layoutCobro.addWidget(self.lblCambio)

        layoutCobro.addStretch()

        self.lblImagenProducto = QLabel("NO IMAGEN")
        self.lblImagenProducto.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.lblImagenProducto.setFont(Estilos.FUENTE_BOLD)
        self.lblImagenProducto.setStyleSheet("color: #3F3F46; border: 1px solid #3F3F46;")
        self.lblImagenProducto.setFixedSize(218, 170)
        layoutCobro.addWidget(self.lblImagenProducto)

        layoutCobro.addSpacing(16)

        btnCobrar = QPushButton("COBRAR")
        btnCobrar.setFont(QFont("SansSerif", 16, QFont.Weight.Bold))
        btnCobrar.setCursor(Qt.CursorShape.PointingHandCursor)
        btnCobrar.setStyleSheet(f"""
            QPushButton {{
                background-color: {Estilos.INDIGO.name()};
                color: white;
                border: none;
                padding: 14px;
            }}
        """)
        btnCobrar.clicked.connect(self.procesarCobro)
        layoutCobro.addWidget(btnCobrar)

        layout.addWidget(panelCobro)

    def actualizarVisibilidadEfectivo(self):
        esEfectivo = self.cmbMetodo.currentText() == "Efectivo"
        self.lblRecibido.setVisible(esEfectivo)
        self.txtRecibido.setVisible(esEfectivo)
        self.lblCambio.setVisible(esEfectivo)
        if not esEfectivo:
            self.txtRecibido.setText("0.00")

    def aplicarDescuento(self):
        val = self.txtDescuento.text().strip()
        try:
            if val.endswith('%'):
                porcentaje = float(val.replace('%', '').strip())
                subtotal = self.ctrl.calcularSubtotal()
                self.ctrl.setDescuento(subtotal * (porcentaje / 100.0))
            else:
                self.ctrl.setDescuento(float(val) if val else 0.0)
        except ValueError:
            self.ctrl.setDescuento(0.0)

    def actualizarSugerencias(self):
        for i in reversed(range(self.sugerenciasLayout.count())): 
            widget = self.sugerenciasLayout.itemAt(i).widget()
            if widget is not None:
                widget.setParent(None)

        q = self.txtBusqueda.text().strip()
        if not q:
            self.sugerenciasArea.setVisible(False)
            return

        res = self.ctrl.buscarProductos(q)
        for p in res:
            btn = QPushButton(f"[{p.getId()}]  {p.getNombre().upper():<22}  ${p.getPrecio():.2f}  stock:{p.getStock():.0f}")
            btn.setFont(Estilos.FUENTE_SMALL)
            btn.setStyleSheet(f"background-color: {Estilos.BG_BLANCO.name()}; color: {Estilos.TEXTO_PRINCIPAL.name()}; text-align: left; padding: 5px; border: none;")
            btn.setCursor(Qt.CursorShape.PointingHandCursor)
            
            def add_to_cart(checked, prod=p):
                self.ctrl.agregarAlCarrito(prod)
                self.mostrarImagen(prod.getImagenRuta())
                self.txtBusqueda.setText("")
                self.sugerenciasArea.setVisible(False)
            
            btn.clicked.connect(add_to_cart)
            self.sugerenciasLayout.addWidget(btn)

        self.sugerenciasArea.setVisible(len(res) > 0)

    def actualizarCambio(self):
        try:
            rec = float(self.txtRecibido.text().strip())
            self.lblCambio.setText(f"Cambio: ${self.ctrl.calcularCambio(rec):.2f}")
        except ValueError:
            pass

    def procesarCobro(self):
        if not self.ctrl.getCarrito():
            QMessageBox.information(self, "Aviso", "El carrito esta vacio")
            return
        
        metodo = self.cmbMetodo.currentText()
        recibido = 0.0
        if metodo == "Efectivo":
            try:
                recibido = float(self.txtRecibido.text().strip())
            except ValueError:
                pass
        else:
            recibido = self.ctrl.calcularTotal()
            
        self.ctrl.procesarCobro(metodo, recibido)

    def refrescarCarrito(self, carrito, total, descuento):
        self.actualizando = True
        self.tablaCarrito.setRowCount(len(carrito))
        for i, item in enumerate(carrito):
            self.tablaCarrito.setItem(i, 0, QTableWidgetItem(item.getProducto().getId()))
            self.tablaCarrito.setItem(i, 1, QTableWidgetItem(item.getProducto().getNombre()))
            
            item2 = QTableWidgetItem(f"${item.getProducto().getPrecio():.2f}")
            item2.setFlags(item2.flags() & ~Qt.ItemFlag.ItemIsEditable)
            self.tablaCarrito.setItem(i, 2, item2)
            
            cantStr = f"{item.getCantidad():.0f}" if item.getProducto().esPorPieza() else f"{item.getCantidad():.2f}"
            self.tablaCarrito.setItem(i, 3, QTableWidgetItem(cantStr))
            
            item4 = QTableWidgetItem(f"${item.getSubtotal():.2f}")
            item4.setFlags(item4.flags() & ~Qt.ItemFlag.ItemIsEditable)
            self.tablaCarrito.setItem(i, 4, item4)
            
            btnQuitar = QPushButton("Quitar")
            btnQuitar.setFont(Estilos.FUENTE_XS)
            btnQuitar.setStyleSheet(f"background-color: {Estilos.ROSE_CLARO.name()}; color: {Estilos.ROSE.name()}; border: none; margin: 2px;")
            btnQuitar.clicked.connect(lambda checked, prod_id=item.getProducto().getId(): self.ctrl.eliminarDelCarrito(prod_id))
            self.tablaCarrito.setCellWidget(i, 5, btnQuitar)
            
        self.actualizando = False
        self.lblTotal.setText(f"${total:.2f}")

    def onCeldaCambiada(self, row, col):
        if self.actualizando or col != 3:
            return
        try:
            prod_id = self.tablaCarrito.item(row, 0).text()
            val = float(self.tablaCarrito.item(row, 3).text().strip())
            if val <= 0:
                self.ctrl.eliminarDelCarrito(prod_id)
            else:
                self.ctrl.setCantidad(prod_id, val)
        except (ValueError, AttributeError):
            pass

    def limpiar(self):
        self.actualizando = True
        self.tablaCarrito.setRowCount(0)
        self.actualizando = False
        self.lblTotal.setText("$0.00")
        self.txtRecibido.setText("0.00")
        self.txtDescuento.setText("0.00")
        self.lblCambio.setText("Cambio: $0.00")
        self.txtBusqueda.setText("")
        self.sugerenciasArea.setVisible(False)
        self.cmbMetodo.setCurrentIndex(0)
        self.actualizarVisibilidadEfectivo()
        self.lblImagenProducto.setPixmap(QPixmap())
        self.lblImagenProducto.setText("")

    def mostrarImagen(self, ruta):
        if not ruta:
            self.lblImagenProducto.setPixmap(QPixmap())
            self.lblImagenProducto.setText("SIN FOTO")
            return
            
        from Util.RutaBase import RutaBase
        if not os.path.isabs(ruta):
            ruta = os.path.join(RutaBase.getRaiz(), ruta)
            
        if os.path.exists(ruta):
            pixmap = QPixmap(ruta)
            if not pixmap.isNull():
                self.lblImagenProducto.setPixmap(pixmap.scaled(218, 170, Qt.AspectRatioMode.KeepAspectRatio, Qt.TransformationMode.SmoothTransformation))
                self.lblImagenProducto.setText("")
            else:
                self.lblImagenProducto.setText("SIN FOTO")
        else:
            self.lblImagenProducto.setText("SIN FOTO")

File: Vista/VentaPanel.cs
using Microsoft.Maui.Controls;
using Microsoft.Maui.Graphics;
using Microsoft.Maui.Layouts;
using System;
using System.Collections.Generic;
using System.Linq;
using punto_de_venta_C_.Modelo;
using punto_de_venta_C_.Controlador;

namespace punto_de_venta_C_.Vista
{
    public class VentaPanel : ContentView
    {
        private readonly VentaControlador ctrl;
        private readonly ConfiguracionTienda config;
        private readonly Usuario cajero;

        private Entry txtBusqueda;
        private VerticalStackLayout sugerenciasPanel;
        private ScrollView sugerenciasScroll;
        private VerticalStackLayout carritoList;
        private Label lblTotal;
        private Entry txtRecibido;
        private Entry txtDescuento;
        private Label lblCambio;
        private Label lblRecibido;
        private Grid containerImagenProducto;
        private Image imgProducto;
        private Label lblPlaceholderImagen;
        private Picker cmbMetodo;
        private bool actualizando = false;

        public VentaPanel(VentaControlador ctrl, ConfiguracionTienda config, Usuario cajero)
        {
            this.ctrl = ctrl;
            this.config = config;
            this.cajero = cajero;
            ctrl.setPanel(this);
            BackgroundColor = Estilos.BG_CLARO;
            construir();
        }

        private void construir()
        {
            var absoluteLayout = new AbsoluteLayout
            {
                WidthRequest = 1000,
                HeightRequest = 700
            };

            var lblBusq = new Label
            {
                Text = "BUSCAR PRODUCTO (nombre o SKU):",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Estilos.TEXTO_TENUE
            };
            AbsoluteLayout.SetLayoutBounds(lblBusq, new Rect(16, 14, 300, 14));
            AbsoluteLayout.SetLayoutFlags(lblBusq, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblBusq);

            txtBusqueda = new Entry
            {
                FontSize = Estilos.FUENTE_SUBTITULO_SIZE,
                BackgroundColor = Estilos.BG_BLANCO,
                TextColor = Estilos.TEXTO_PRINCIPAL
            };
            txtBusqueda.TextChanged += (s, e) => actualizarSugerencias();
            AbsoluteLayout.SetLayoutBounds(txtBusqueda, new Rect(16, 32, 700, 44));
            AbsoluteLayout.SetLayoutFlags(txtBusqueda, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(txtBusqueda);

            sugerenciasPanel = new VerticalStackLayout
            {
                BackgroundColor = Estilos.BG_BLANCO
            };

            sugerenciasScroll = new ScrollView
            {
                Content = sugerenciasPanel,
                IsVisible = false,
                BackgroundColor = Estilos.BG_BLANCO
            };
            AbsoluteLayout.SetLayoutBounds(sugerenciasScroll, new Rect(16, 76, 700, 200));
            AbsoluteLayout.SetLayoutFlags(sugerenciasScroll, AbsoluteLayoutFlags.None);
            sugerenciasScroll.ZIndex = 10;

            var lblDetalle = new Label
            {
                Text = "DETALLE DE VENTA",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Estilos.TEXTO_TENUE
            };
            AbsoluteLayout.SetLayoutBounds(lblDetalle, new Rect(16, 84, 200, 14));
            AbsoluteLayout.SetLayoutFlags(lblDetalle, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(lblDetalle);

            carritoList = new VerticalStackLayout();
            var scrollCarrito = new ScrollView
            {
                Content = carritoList,
                BackgroundColor = Estilos.BG_BLANCO
            };
            AbsoluteLayout.SetLayoutBounds(scrollCarrito, new Rect(16, 100, 700, 480));
            AbsoluteLayout.SetLayoutFlags(scrollCarrito, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(scrollCarrito);

            var panelCobro = new AbsoluteLayout
            {
                BackgroundColor = Estilos.BG_OSCURO,
                WidthRequest = 250,
                HeightRequest = 600
            };
            AbsoluteLayout.SetLayoutBounds(panelCobro, new Rect(726, 14, 250, 600));
            AbsoluteLayout.SetLayoutFlags(panelCobro, AbsoluteLayoutFlags.None);
            absoluteLayout.Add(panelCobro);

            var lblResumen = new Label
            {
                Text = "RESUMEN",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Color.FromRgb(129, 140, 248)
            };
            AbsoluteLayout.SetLayoutBounds(lblResumen, new Rect(16, 16, 218, 14));
            AbsoluteLayout.SetLayoutFlags(lblResumen, AbsoluteLayoutFlags.None);
            panelCobro.Add(lblResumen);

            containerImagenProducto = new Grid
            {
                WidthRequest = 218,
                HeightRequest = 170,
                BackgroundColor = Color.FromRgb(30, 30, 32)
            };
            imgProducto = new Image { Aspect = Aspect.AspectFit };
            lblPlaceholderImagen = new Label
            {
                Text = "NO IMAGEN",
                HorizontalTextAlignment = TextAlignment.Center,
                VerticalTextAlignment = TextAlignment.Center,
                TextColor = Color.FromRgb(113, 113, 122)
            };
            containerImagenProducto.Add(lblPlaceholderImagen);
            containerImagenProducto.Add(imgProducto);
            AbsoluteLayout.SetLayoutBounds(containerImagenProducto, new Rect(16, 340, 218, 170));
            AbsoluteLayout.SetLayoutFlags(containerImagenProducto, AbsoluteLayoutFlags.None);
            panelCobro.Add(containerImagenProducto);

            lblTotal = new Label
            {
                Text = "$0.00",
                FontSize = 38,
                FontAttributes = FontAttributes.Bold,
                TextColor = Color.FromRgb(165, 180, 252)
            };
            AbsoluteLayout.SetLayoutBounds(lblTotal, new Rect(16, 34, 218, 46));
            AbsoluteLayout.SetLayoutFlags(lblTotal, AbsoluteLayoutFlags.None);
            panelCobro.Add(lblTotal);

            var lblDescuento = new Label
            {
                Text = "DESCUENTO ($ o %)",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Color.FromRgb(113, 113, 122)
            };
            AbsoluteLayout.SetLayoutBounds(lblDescuento, new Rect(16, 90, 218, 14));
            AbsoluteLayout.SetLayoutFlags(lblDescuento, AbsoluteLayoutFlags.None);
            panelCobro.Add(lblDescuento);

            txtDescuento = new Entry
            {
                Text = "0.00",
                FontSize = Estilos.FUENTE_BOLD_SIZE,
                BackgroundColor = Color.FromRgb(39, 39, 42),
                TextColor = Color.FromRgb(52, 211, 153)
            };
            txtDescuento.TextChanged += (s, e) => aplicarDescuento();
            AbsoluteLayout.SetLayoutBounds(txtDescuento, new Rect(16, 108, 218, 36));
            AbsoluteLayout.SetLayoutFlags(txtDescuento, AbsoluteLayoutFlags.None);
            panelCobro.Add(txtDescuento);

            var lblMetodoLbl = new Label
            {
                Text = "METODO DE PAGO",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Color.FromRgb(113, 113, 122)
            };
            AbsoluteLayout.SetLayoutBounds(lblMetodoLbl, new Rect(16, 158, 218, 14));
            AbsoluteLayout.SetLayoutFlags(lblMetodoLbl, AbsoluteLayoutFlags.None);
            panelCobro.Add(lblMetodoLbl);

            cmbMetodo = new Picker
            {
                ItemsSource = new List<string> { "Efectivo", "Tarjeta" },
                SelectedIndex = 0,
                FontSize = Estilos.FUENTE_BOLD_SIZE
            };
            cmbMetodo.SelectedIndexChanged += (s, e) => actualizarVisibilidadEfectivo();
            AbsoluteLayout.SetLayoutBounds(cmbMetodo, new Rect(16, 176, 218, 36));
            AbsoluteLayout.SetLayoutFlags(cmbMetodo, AbsoluteLayoutFlags.None);
            panelCobro.Add(cmbMetodo);

            lblRecibido = new Label
            {
                Text = "IMPORTE RECIBIDO",
                FontSize = Estilos.FUENTE_XS_SIZE,
                TextColor = Color.FromRgb(113, 113, 122)
            };
            AbsoluteLayout.SetLayoutBounds(lblRecibido, new Rect(16, 224, 218, 14));
            AbsoluteLayout.SetLayoutFlags(lblRecibido, AbsoluteLayoutFlags.None);
            panelCobro.Add(lblRecibido);

            txtRecibido = new Entry
            {
                Text = "0.00",
                FontSize = 22,
                FontAttributes = FontAttributes.Bold,
                BackgroundColor = Color.FromRgb(39, 39, 42),
                TextColor = Colors.White
            };
            txtRecibido.TextChanged += (s, e) => actualizarCambio();
            AbsoluteLayout.SetLayoutBounds(txtRecibido, new Rect(16, 242, 218, 48));
            AbsoluteLayout.SetLayoutFlags(txtRecibido, AbsoluteLayoutFlags.None);
            panelCobro.Add(txtRecibido);

            lblCambio = new Label
            {
                Text = "Cambio: $0.00",
                FontSize = Estilos.FUENTE_BOLD_SIZE,
                TextColor = Color.FromRgb(52, 211, 153)
            };
            AbsoluteLayout.SetLayoutBounds(lblCambio, new Rect(16, 300, 218, 22));
            AbsoluteLayout.SetLayoutFlags(lblCambio, AbsoluteLayoutFlags.None);
            panelCobro.Add(lblCambio);

            var btnCobrar = new Button
            {
                Text = "COBRAR",
                FontSize = 16,
                FontAttributes = FontAttributes.Bold,
                BackgroundColor = Estilos.INDIGO,
                TextColor = Colors.White
            };
            btnCobrar.Clicked += (s, e) => procesarCobro();
            AbsoluteLayout.SetLayoutBounds(btnCobrar, new Rect(16, 536, 218, 52));
            AbsoluteLayout.SetLayoutFlags(btnCobrar, AbsoluteLayoutFlags.None);
            panelCobro.Add(btnCobrar);

            absoluteLayout.Add(sugerenciasScroll);

            Content = absoluteLayout;
        }

        private void actualizarVisibilidadEfectivo()
        {
            bool esEfectivo = "Efectivo".Equals(cmbMetodo.SelectedItem);
            lblRecibido.IsVisible = esEfectivo;
            txtRecibido.IsVisible = esEfectivo;
            lblCambio.IsVisible = esEfectivo;
            if (!esEfectivo)
                txtRecibido.Text = "0.00";
        }

        private void aplicarDescuento()
        {
            try
            {
                string val = txtDescuento.Text?.Trim() ?? "";
                if (val.EndsWith("%"))
                {
                    double porcentaje = double.Parse(val.Replace("%", "").Trim());
                    double subtotal = ctrl.calcularSubtotal();
                    ctrl.setDescuento(subtotal * (porcentaje / 100.0));
                }
                else
                {
                    ctrl.setDescuento(double.Parse(val));
                }
            }
            catch (Exception)
            {
                ctrl.setDescuento(0);
            }
        }

        private void actualizarSugerencias()
        {
            sugerenciasPanel.Children.Clear();
            string q = txtBusqueda.Text?.Trim() ?? "";
            if (string.IsNullOrWhiteSpace(q))
            {
                sugerenciasScroll.IsVisible = false;
                return;
            }
            List<Producto> res = ctrl.buscarProductos(q);
            foreach (Producto p in res)
            {
                var btn = new Button
                {
                    Text = string.Format("[{0}]  {1}  ${2:0.00}  stock:{3}", p.getId(), p.getNombre().ToUpper(), p.getPrecio(), p.getStock()),
                    FontSize = Estilos.FUENTE_SMALL_SIZE,
                    BackgroundColor = Estilos.BG_BLANCO,
                    TextColor = Estilos.TEXTO_PRINCIPAL,
                    HorizontalOptions = LayoutOptions.Fill,
                    CornerRadius = 0
                };
                btn.Clicked += (s, e) =>
                {
                    ctrl.agregarAlCarrito(p);
                    mostrarImagen(p.getImagenRuta());
                    txtBusqueda.Text = "";
                    sugerenciasPanel.Children.Clear();
                    sugerenciasScroll.IsVisible = false;
                };
                sugerenciasPanel.Children.Add(btn);
            }
            sugerenciasScroll.IsVisible = res.Count > 0;
        }

        private void actualizarCambio()
        {
            try
            {
                double rec = double.Parse(txtRecibido.Text?.Trim() ?? "0");
                lblCambio.Text = string.Format("Cambio: ${0:0.00}", ctrl.calcularCambio(rec));
            }
            catch (Exception)
            {
            }
        }

        private void procesarCobro()
        {
            if (ctrl.getCarrito().Count == 0)
            {
                if (Application.Current?.MainPage != null)
                {
                    Application.Current.MainPage.DisplayAlert("Aviso", "El carrito esta vacio", "OK");
                }
                return;
            }
            string metodo = cmbMetodo.SelectedItem?.ToString() ?? "Efectivo";
            double recibido = 0;
            if ("Efectivo".Equals(metodo))
            {
                try
                {
                    recibido = double.Parse(txtRecibido.Text?.Trim() ?? "0");
                }
                catch (Exception)
                {
                }
            }
            else
            {
                recibido = ctrl.calcularTotal();
            }
            ctrl.procesarCobro(metodo, recibido);
        }

        public void refrescarCarrito(List<ItemCarrito> carrito, double total, double descuento)
        {
            actualizando = true;
            carritoList.Children.Clear();

            var header = new Grid
            {
                BackgroundColor = Estilos.BG_ZINC_100,
                HeightRequest = 35,
                ColumnDefinitions = new ColumnDefinitionCollection
                {
                    new ColumnDefinition { Width = GridLength.Star },
                    new ColumnDefinition { Width = 100 },
                    new ColumnDefinition { Width = 80 },
                    new ColumnDefinition { Width = 100 },
                    new ColumnDefinition { Width = 80 }
                }
            };
            header.Add(new Label { Text = "Producto", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center, Margin = new Thickness(10, 0) }, 0);
            header.Add(new Label { Text = "Precio Unit", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 1);
            header.Add(new Label { Text = "Cant", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 2);
            header.Add(new Label { Text = "Subtotal", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 3);
            header.Add(new Label { Text = "Quitar", FontAttributes = FontAttributes.Bold, VerticalOptions = LayoutOptions.Center }, 4);
            carritoList.Children.Add(header);

            foreach (ItemCarrito item in carrito)
            {
                var rowGrid = new Grid
                {
                    HeightRequest = 44,
                    ColumnDefinitions = new ColumnDefinitionCollection
                    {
                        new ColumnDefinition { Width = GridLength.Star },
                        new ColumnDefinition { Width = 100 },
                        new ColumnDefinition { Width = 80 },
                        new ColumnDefinition { Width = 100 },
                        new ColumnDefinition { Width = 80 }
                    }
                };

                rowGrid.Add(new Label { Text = item.getProducto().getNombre(), VerticalOptions = LayoutOptions.Center, Margin = new Thickness(10, 0) }, 0);
                rowGrid.Add(new Label { Text = string.Format("${0:0.00}", item.getProducto().getPrecio()), VerticalOptions = LayoutOptions.Center }, 1);

                var cantEntry = new Entry
                {
                    Text = item.getProducto().esPorPieza() ? string.Format("{0:0}", item.getCantidad()) : string.Format("{0:0.00}", item.getCantidad()),
                    VerticalOptions = LayoutOptions.Center
                };
                string currentId = item.getProducto().getId();
                cantEntry.Completed += (s, e) =>
                {
                    try
                    {
                        double val = double.Parse(cantEntry.Text);
                        if (val <= 0)
                            ctrl.eliminarDelCarrito(currentId);
                        else
                            ctrl.setCantidad(currentId, val);
                    }
                    catch (Exception)
                    {
                    }
                };
                rowGrid.Add(cantEntry, 2);

                rowGrid.Add(new Label { Text = string.Format("${0:0.00}", item.getSubtotal()), VerticalOptions = LayoutOptions.Center }, 3);

                var btnQuitar = new Button
                {
                    Text = "Quitar",
                    FontSize = Estilos.FUENTE_XS_SIZE,
                    TextColor = Estilos.ROSE,
                    BackgroundColor = Estilos.ROSE_CLARO,
                    HeightRequest = 30,
                    Padding = new Thickness(0)
                };
                btnQuitar.Clicked += (s, e) => ctrl.eliminarDelCarrito(currentId);
                rowGrid.Add(btnQuitar, 4);

                carritoList.Children.Add(rowGrid);
            }
            actualizando = false;
            lblTotal.Text = string.Format("${0:0.00}", total);
        }

        public void limpiar()
        {
            actualizando = true;
            carritoList.Children.Clear();
            actualizando = false;
            lblTotal.Text = "$0.00";
            txtRecibido.Text = "0.00";
            txtDescuento.Text = "0.00";
            lblCambio.Text = "Cambio: $0.00";
            txtBusqueda.Text = "";
            sugerenciasScroll.IsVisible = false;
            cmbMetodo.SelectedIndex = 0;
            actualizarVisibilidadEfectivo();
            imgProducto.Source = null;
            lblPlaceholderImagen.Text = "NO IMAGEN";
            lblPlaceholderImagen.IsVisible = true;
        }

        private void mostrarImagen(string ruta)
        {
            var src = Util.RutaBase.getImagenSource(ruta);
            if (src != null)
            {
                imgProducto.Source = src;
                lblPlaceholderImagen.IsVisible = false;
            }
            else
            {
                imgProducto.Source = null;
                lblPlaceholderImagen.Text = "SIN FOTO";
                lblPlaceholderImagen.IsVisible = true;
            }
        }
    }
}
